base/compiler/ssair/passes.jl

```diff
@@ -1106,24 +1106,24 @@ function sroa_pass!(ir::IRCode, inlining::Union{Nothing,InliningState}=nothing)
             val = stmt.args[2]
         end
         struct_typ = widenconst(argextype(val, compact))
-        struct_typ_unwrapped = unwrap_unionall(struct_typ)
-        if isa(struct_typ, Union)
-            struct_typ_unwrapped = unswitchtypeunion(struct_typ_unwrapped)
-        end
-        if isa(struct_typ_unwrapped, Union) && is_isdefined
-            lift_comparison!(isdefined, compact, idx, stmt, lifting_cache, 𝕃ₒ)
+        struct_typ_name = argument_datatype(struct_typ)
+        if struct_typ_name === nothing
+            if isa(struct_typ, Union)
+                lift_comparison!(isdefined, compact, idx, stmt, lifting_cache, 𝕃ₒ)
+            end
             continue
+        else
+            struct_typ_name = struct_typ_name.name
         end
-        isa(struct_typ_unwrapped, DataType) || continue
 
-        struct_typ_unwrapped.name.atomicfields == C_NULL || continue # TODO: handle more
+        struct_typ_name.atomicfields == C_NULL || continue # TODO: handle more
         if !((field_ordering === :unspecified) ||
              (field_ordering isa Const && field_ordering.val === :not_atomic))
             continue
         end
 
         # analyze this mutable struct here for the later pass
-        if ismutabletype(struct_typ_unwrapped)
+        if struct_typ_name.flags & 0x2 == 0x2 # if this struct is mutable
```

Comment:
`ismutabletype(struct_typ_unwrapped.wrapper)`

The flags field is very unstable, so it is a bad idea to hard code assumptions about it.

Reply:
`struct_typ_name.wrapper` is `UnionAll` so it is not applicable to `ismutabletype`. We probably want to define `ismutabletypename`?